### PR TITLE
Inpainting start

### PR DESCRIPTION
**Pull request by Akrielz**
This PR adds the capability of using a Pre-trained Diffusion Model such that it can apply inpainitng, without needing of re-training the model.

The technique is highlighted in this paper:
https://arxiv.org/abs/2201.09865

The capability of inpainting is captured in bin/correct_structure.py

The process goes as following: reads pdb -> extracts angles -> inpaint missing info -> use nerf to go from angles to coordinates -> combine old structure elements with the new one -> write it the new pdb.

The code also modified nerf to allow to use the old set of coordinates if given.

TODO List:
- [x] Add inpaiting for masked zone
- [x] Make sure the results from the inpainting are good enough
- [x] Add capability to combine the new generated structure with the old structure
- [x] Make the script work on proteins longer than 128
- [x] Add capability to detect the missing residues
- [x] Make the script work with PDBs that are missing actual information
- [ ] Make the script iterate over all the chains
- [x] Make the script evaluate all the generated version, and automatically decide which is best
- [ ] Use their tools to record a diffusion process
- [x] Make the script accept more PDBs at once
- [x] Add documentation
- [x] Fine tune the backbone with physical constraints
- [x] Run on all the broken structures
- [x] Add header with sequences to structures
- [x] Run on all the broken structures with header 
- [ ] Fix mini bug for starting structure inside NERF.

### CODE REVIEW

**Comment by jtourjeman** [comment]
@Akrielz Should this be linked somewhere in the Scaffold-constrained binding project?

**Comment by jtourjeman** [comment]
Also this PR should likely have _some_ information in it about is actually happening, yeah? 😉 @Akrielz

**Comment by Akrielz** [comment]
@jtourjeman Yeah, I will link it properly and add description tomorrow to make everything smother.

File: bin/structure_utils.py [hunk 1]
```diff
@@ -18,15 +24,277 @@
     'ALA': 'A', 'VAL': 'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'
 }
 
+d1to3 = {v: k for k, v in d3to1.items()}
+
+d3to1_extended = {
```
Comment (arolihas) [inline bin/structure_utils.py:29]: Can you explain how you came up with this mapping? It seems like a lot of these codes are for hetero-atoms or ligands (basically not a part of the protein)

Comment (arolihas) [reply bin/structure_utils.py:29]: Spoke with Jeff, they actually are a part of biopython https://github.com/biopython/biopython/blob/257143be9196b77619d3d8cadc22039212681e08/Bio/Data/PDBData.py

File: bin/structure_utils.py [hunk 2]
```diff
@@ -139,68 +630,229 @@ def get_sequence_from_header(header: List[str]) -> Optional[Dict[str, str]]:
...
+    -----
+    The sequence is defined as the residues in the structure.
+    """
+
     chains = get_chains(structure)
     structure = filter_only_ca(structure)
     return {
         chain: "".join(
-            d3to1.get(res_name, "?")
+            d3to1_all.get(res_name, "?")
```
Comment (arolihas) [inline bin/structure_utils.py:655]: How do we deal with "?" residues? Do they end up in the PDB file?

Comment (Akrielz) [reply bin/structure_utils.py:655]: Those are marked with "UNK" in the final PDB file.

Comment (Akrielz) [reply bin/structure_utils.py:655]: But this is the case only when we had it "unknown" in the header and it was also missing from the structure.

Comment (arolihas) [reply bin/structure_utils.py:653]: And If they are marked with UNK they should show up as X, got it.

**Pull request merged**